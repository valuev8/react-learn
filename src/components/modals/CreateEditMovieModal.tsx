import React, {FC, FormEvent, useState} from 'react';
import FormGroup from '../../shared/components/formGroup/formGroup';
import {Movie} from '../../shared/models/movie.type';
import Button from '../../shared/components/button/Button';
import ModalWindow from '../../shared/components/modal/ModalWindow';
import useModal from '../../shared/hooks/useModal';
import {Formik, FormikProps} from 'formik';
import {movieFormValidator} from '../../shared/validation/movie-form.validation';

type fieldConfig = {
  label: string,
  placeholder?: string,
  readOnly?: boolean,
  type?: 'date' | 'select' | 'text' | 'number';
}

type ModalProps = {
  data?: Movie,
  type?: string;
  onConfirm?: (movie: Partial<Movie> | Movie) => void;
}

const fieldConfig: { [key: string]: fieldConfig } = {
  title: {placeholder: 'Title', type: 'text', label: 'title'},
  release_date: {placeholder: 'Select Date', type: 'date', label: 'release date'},
  poster_path: {placeholder: 'Movie URL here', type: 'text', label: 'movie url'},
  overview: {placeholder: 'Overview here', type: 'text', label: 'overview'},
  genres: {placeholder: 'Select genre', type: 'select', label: 'genre'},
  runtime: {placeholder: 'Runtime here', type: 'number', label: 'runtime'},
  vote_average: {placeholder: 'Rating', type: 'number', label: 'rating'},
};

const defaultMovie: Partial<Movie> = {
  title: '',
  release_date: '',
  poster_path: '',
  overview: '',
  genres: [],
  runtime: '' as any,
  vote_average: '' as any,
};

const CreateEditMovieModal: FC<ModalProps> = (props) => {
  const [movie] = useState(props.data || defaultMovie);
  const {isShowing, toggle} = useModal();

  const handleConfirm = (movie: Movie | Partial<Movie>) => {
    props.onConfirm({
      ...movie,
      runtime: Number(movie.runtime),
      vote_average: Number(movie.vote_average),
    });
    toggle();
  };

  // TODO: move form to separate component //
  return (
    <React.Fragment>
      <Button
        width='100'
        onClick={ toggle }>
        { props.data ? 'Edit' : 'Add Movie' }
      </Button>
      <ModalWindow
        isShowing={ isShowing }
        hide={ toggle }
        title={ props.data ? 'Edit Movie' : 'Add Movie' }>
        <Formik
          validate={ movieFormValidator }
          initialValues={ movie }
          onSubmit={ handleConfirm }>
          {(props: FormikProps<Movie>) => (
            <form onSubmit={(e: FormEvent) => {
              e.preventDefault();
              props.handleSubmit();
            }}>
              { Object.keys(fieldConfig).map((key: string) =>
                <FormGroup
                  key={ key }
                  name={ key }
                  label={ fieldConfig[key].label }
                  type={ fieldConfig[key].type }
                  placeholder={ fieldConfig[key].placeholder } />,
              )}
              <div className="modal-footer">
                <Button onClick={ toggle }> Close </Button>
                <Button
                  type='submit'
                  theme='success'> Confirm </Button>
              </div>
            </form>
          )}
        </Formik>
      </ModalWindow>
    </React.Fragment>
  );
};

export default CreateEditMovieModal;
