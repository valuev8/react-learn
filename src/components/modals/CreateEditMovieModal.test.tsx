import React from 'react';
import CreateEditMovieModal from './CreateEditMovieModal';
import { render } from '@testing-library/react';
import { Movie } from '../../shared/models/movie.type';

const mockData = {
  title: 'test'
} as Movie;

describe('CreateEditMovieModal', () => {
  it('Should render', () => {
    const { asFragment } = render(<CreateEditMovieModal type='create' onConfirm={() => {}}/>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should have Edit titles if data is provided', () => {
    const { getByText } = render(<CreateEditMovieModal data={mockData} type='create' onConfirm={() => {}}/>);
    const element = getByText(/edit/i);

    expect(element).toBeTruthy();
  });

  it('Should have Add Movie titles if data is provided', () => {
    const { getByText } = render(<CreateEditMovieModal type='create' onConfirm={() => {}}/>);
    const element = getByText(/add movie/i);

    expect(element).toBeTruthy();
  });
});
