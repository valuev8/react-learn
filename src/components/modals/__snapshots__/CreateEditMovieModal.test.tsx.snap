// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditMovieModal Should render 1`] = `
<DocumentFragment>
  <button
    class="sc-AxhCb eOHJTW"
    type="button"
    width="100"
  >
    Add Movie
  </button>
</DocumentFragment>
`;
